Adds non-fire melting

Domic2 opened the pull request:
WIP, slapped-together testing version.

Works surprisingly well. We can bypass constant performance issues by only doing the necessary checks when a zone detects higher-than-usual temperatures. May start dying if there are too many turfs doing checks, although this doesn't seem to happen easily.

Some meltables not implemented, others are implemented badly and/or partially.

It seems like the Torch is extremely vulnerable to melting (lots of non-fireshutter windows and steel walls), mapping changes should be done alongside this, assuming it gets finished and is considered acceptable.

Would definitely like to hear feedback on this, I'm not really familiar with atmos code, and I'm not sure if the check should happen during tick(), or if it should be somewhere else.

Domic2 committed:
Adds non-fire melting

EXPERIMENTAL
(for now)

Tennessee116 commented on code/modules/ZAS/Turf.dm:
```diff
@@ -289,3 +289,152 @@
 	air.copy_from(zone.air)
 	air.group_multiplier = 1
+
+//For meltables
+obj/proc/can_melt(temp)
+	return 0
+
+/obj/machinery/portable_atmospherics/canister/can_melt(temp)
+	if(temp > temperature_resistance)
+		return 1
+	return 0
+
+/obj/structure/grille/can_melt(temp)
+	if(temp > material.melting_point)
+		return 1
+	return 0
+
+/obj/structure/wall_frame/can_melt(temp)
+	if(temp > material.melting_point)
+		return 1
+	return 0
+
+/obj/structure/girder/can_melt(temp)
+	if(temp > SSmaterials.get_material_by_name(MATERIAL_STEEL).melting_point)
+		return 1
+	return 0
+
+/obj/machinery/door/firedoor/can_melt(temp)
+	if (temp > (SSmaterials.get_material_by_name(MATERIAL_STEEL).melting_point + 400))
+		return 1
+	return 0
```
I suggest using TRUE/FALSE instead of 1/0.

Spookerton replied:
And I require it!

Domic2 replied:
Roger that, thanks

MuckerMayhem commented on code/modules/ZAS/Turf.dm:
```diff
@@ -289,3 +289,152 @@
+	if (zone.air.temperature > SSmaterials.get_material_by_name(MATERIAL_STEEL).melting_point && !is_plating())
+		meltable_turfs |= src
+
+	var/turf/cardinal = CardinalTurfs(FALSE)
```
What's the purpose of checking nearby walls for being meltable? Won't the zone check all its turfs as it is? And if the temperatures rises in a neighboring zone, won't they also check for those same turfs anyway if they are in a different zone? Seems like this could cause a lot of unnecessary overlapping, if I'm understanding what you're doing here correctly.

Domic2 replied:
Unfortunately, unless I am mistaken, zones do not keep track of walls. The overlap is primarily there to make walls which are surrounded by hot turfs melt before those which are only affected from one side, as they should be taking in more energy and melting faster

MuckerMayhem replied:
Makes sense, though I do fear for performance should big fires happen. Also, as an aside, it may be wise to make sure that planet fires don't absolutely demolish the server's performance as well.

Domic2 replied:
... I thought planet turfs were unsimulated... Damn.

Domic2 replied:
I still have to test how many turfs we can actually handle, but the len check is able to deny planets just fine

Tennessee116 commented on code/modules/ZAS/Turf.dm:
```diff
@@ -289,3 +289,152 @@
+
+/obj/structure/window/heat_damage(temp)
+	for(var/obj/machinery/door/protector in loc.contents)
+		if (istype(protector, /obj/machinery/door/blast) || istype(protector, /obj/machinery/door/firedoor))
```
What about regular shutters?

Domic2 replied:
Uh, regular shutters? Like the ones in engineering? They're an /obj/machinery/door/blast subtype, so should be good

Tennessee116 commented on code/modules/ZAS/Turf.dm:
```diff
@@ -289,3 +289,152 @@
+	if (temp > material.melting_point)
+		take_damage(log(RAND_F(0.9, 1.1) * (temp - material.melting_point)))
+
+/obj/structure/girder/heat_damage(temp)
+	var/material/default_material = SSmaterials.get_material_by_name(MATERIAL_STEEL)
+	if (temp > default_material.melting_point)
+		take_damage(log(RAND_F(0.9, 1.1) * (temp - default_material.melting_point)))
+
+/obj/structure/grille/heat_damage(temp)
+	var/material/default_material = SSmaterials.get_material_by_name(MATERIAL_STEEL)
+	if (temp > default_material.melting_point)
+		take_damage(log(RAND_F(0.9, 1.1) * (temp - default_material.melting_point)))
```
Shouldn't there just be an `/obj/structure/heat_damage` since these are both the same proc?

Domic2 replied:
Yea, probably, was gonna get to it today

Domic2 replied:
All done!

Tennessee116 commented:
I also suggest that `type/` should be `/type/` (`atom/ `-> `/atom/`, `turf/` -> `/turf/`).

Domic2 committed:
Make it better

We now check for how many contents a zone has, to avoid processing entire planets.
Structures now have a more general setup, allowing for most structures with defined materials (and proper damage procs) to be meltable.

SierraKomodo commented on code/modules/ZAS/Turf.dm:
```diff
@@ -289,3 +289,149 @@
+		return TRUE
+	return FALSE
+
+/obj/structure/girder/can_melt(temp)
+	if (!material)// they usually don't have those
+		material = SSmaterials.get_material_by_name(MATERIAL_STEEL)
+	. = ..()
+
+/obj/structure/grille/can_melt(temp)
+	if (!material)
+		material = SSmaterials.get_material_by_name(MATERIAL_STEEL)
+	. = ..()
```
I'd prefer these being given default materials that they always initialize with instead.

SierraKomodo commented on code/modules/ZAS/Turf.dm:
```diff
@@ -289,3 +289,149 @@
+//TURFS
+/turf/simulated/floor/heat_damage(temp)
+	var/material/default_material = SSmaterials.get_material_by_name(MATERIAL_STEEL)
+	if (temp > default_material.melting_point)
```
The temperature checks here seem redundant, since `can_melt()` already checks this and looks to be a prerequisite. If you do need to include checks, I'd suggest `if (can_melt(temp))` instead.

SierraKomodo commented on code/modules/ZAS/Turf.dm:
```diff
@@ -289,3 +289,149 @@
+
+/turf/simulated/wall/heat_damage(temp)
+	burn(temp)
+	if (temp > material.melting_point)
```
See above

SierraKomodo commented on code/modules/ZAS/Turf.dm:
```diff
@@ -289,3 +289,149 @@
+
+//STRUCTURES
+/obj/structure/heat_damage(temp)//we don't need to check for materials, can_melt should be called first, it should handle it
+	if (temp > (material.melting_point))
```
See above

SierraKomodo commented on code/modules/ZAS/Turf.dm:
```diff
@@ -289,3 +289,149 @@
+	var/melting_point = material.melting_point
+	if (reinf_material)
+		melting_point += 0.25*reinf_material.melting_point
+	if (temp > melting_point)
```
See above

SierraKomodo commented on code/modules/ZAS/Turf.dm:
```diff
@@ -289,3 +289,149 @@
+
+//MACHINERY
+/obj/machinery/portable_atmospherics/canister/heat_damage(temp)
+	if(temp > temperature_resistance)
```
See above

SierraKomodo commented on code/modules/ZAS/Turf.dm:
```diff
@@ -289,3 +289,149 @@
+
+/obj/machinery/door/firedoor/heat_damage(temp)
+	var/material/default_material = SSmaterials.get_material_by_name(MATERIAL_STEEL)
+	if (temp > (default_material.melting_point) + 400)
```
See above

SierraKomodo commented on code/modules/ZAS/Turf.dm:
```diff
@@ -289,3 +289,149 @@
+	var/list/meltable_turfs = list()
+	var/list/meltable_nonturfs = list()
+
+//called by the zone every time the air contents change AND the temperature is higher than usual, we check for turfs here and determine if we need to start processing
```
https://github.com/Baystation12/Baystation12/wiki/Guide-to-SpacemanDMM-Docblocks

```suggestion
/// Called by the zone every time the air contents change AND the temperature is higher than usual, we check for turfs here and determine if we need to start processing
```

SierraKomodo reviewed (changes requested):
Not thorough.

Domic2 commented:
I've been testing this for a while now, there's definitely better ways to do this, but they are a bit complicated, so this will take a while

SierraKomodo commented:
Is this still being worked on?

Domic2 commented:
Yup, did a few rewrites and had a college break, slowly getting back into it now

Spookerton commented:
Can reopen if continued.
